core/src/Streamly/Internal/Data/Stream/TH.hs

```diff
@@ -0,0 +1,274 @@
+-- @
+--
+-- Create a type with a concurrent zipping applicative property:
+--
+-- The following should be in scope: Stream.repeat, Concur.MonadAsync
+--
+-- @
+-- import qualified Streamly.Data.Stream as Stream
+-- import qualified Streamly.Internal.Data.Stream.Concurrent as Concur
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+-- $(StreamTH.makeStreamZipApplicative "ZipConcurrent" "Concur.zipWith" True)
+-- @
+--
+makeStreamZipApplicative
+    :: String -- ^ Name of the type
+    -> String -- ^ Function to use for (<*>)
+    -> Bool   -- ^ 'True' if (<*>) has concurrent properties
+    -> Q [Dec]
+makeStreamZipApplicative dtNameStr apOpStr isConcurrent = do
+    typeDec <- makeTypeDec dtNameStr
+    functorI <- makeStreamFunctor dtNameStr
+    applicativeI <-
+        mkBinaryOpInstance
+            dtNameStr
+            "Applicative"
+            classConstraints
+            ("pure", "Stream.repeat")
+            ("(<*>)", apOpStr)
+    return [typeDec, functorI, applicativeI]
+
+    where
+
+    classConstraints =
+        if isConcurrent
+        then ["Concur.MonadAsync"]
+        else ["Monad"]
+
+-- XXX This would become concatMapUsing
+-- | Create a type with specific stream combination properties.
+--
+-- Create a type with serial interleave properties:
+--
+-- The following should be in scope: Stream.fromPure
+--
+-- @
+-- import qualified Streamly.Data.Stream as Stream
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+-- bindInterleave = flip (Stream.concatMapWith Stream.interleave)
+-- $(StreamTH.makeStreamMonad "InterleaveSerial" "bindInterleave" False)
+-- @
+--
+-- Create a type with eager concurrency properties:
+--
+-- The following should be in scope: Stream.fromPure, Concur.MonadAsync
+--
+-- @
+-- import qualified Streamly.Data.Stream as Stream
+-- import qualified Streamly.Internal.Data.Stream.Concurrent as Concur
+-- import qualified Streamly.Internal.Data.Stream.TH as StreamTH
+-- bindConcurrent = flip (Concur.concatMapWith (Concur.eager True))
+-- $(StreamTH.makeStreamMonad "Parallel" "bindConcurrent" True)
+-- @
+--
+makeStreamMonad
+    :: String -- ^ Name of the type
+    -> String -- ^ Function to use for (>>=)
+    -> Bool   -- ^ 'True' if (>>=) has concurrent properties
+    -> Q [Dec]
+makeStreamMonad dtNameStr bindOpStr isConcurrent = do
+    typeDec <- makeTypeDec dtNameStr
+    functorI <- makeStreamFunctor dtNameStr
+    monadI <-
+        mkBinaryOpInstance
+            dtNameStr
+            "Stream"
+            classConstraints
+            ("return", "Stream.fromPure")
+            ("(>>=)", bindOpStr)
+    applicativeI <-
+        mkBinaryOpInstance
+            dtNameStr
+            "Applicative"
+            classConstraints
+            ("pure", "return")
+            ("(<*>)", "ap")
+    return [typeDec, functorI, monadI, applicativeI]
+
+    where
+
+    classConstraints =
+        if isConcurrent
+        then ["Concur.MonadAsync"]
+        else ["Monad"]
```

Comment:
This can also be fully qualified I guess.

Reply:
Or maybe we can assume Prelude as implicitly imported.

Reply:
Do you mean `Prelude.Monad`?

Reply:
We can just make it qualified.

Reply:
Lot of things we use from `Prelude`. We can assume an implicit import. We can churn this later if required.